## Defining the question

### Specifying the data analytical question

Implement a naives bayes Classifier model, conduct different experiments, and calculate the resulting metrics

### Defining the Metric of success

Following performance metrics are used to analyze the 
performance of the Naive bayes classifier models with different metrics.

1. **Precision**: It is used to measures the classifier exactness. It is 
also known as positive predictive value (PPV). 

2. **Recall**: It measures the classifier completeness. It is also known 
as sensitivity or hit rate, or true positive rate (TPR). 
Specificity: It measures the proportion of actual negatives that 
are correctly identified and also called a true negative rate 
(TNR). 

3. **Accuracy**: It measures the fraction of predictions our model got 
right. It can be calculated as the total number of correct 
predictions divided by the total number of test cases. 

4. **F1-Score**: It measures the balance between precision and recall. 


### Understanding the context

Naive Bayes uses a similar method to predict the probability of different class based on various attributes. This algorithm is mostly used in text classification and with problems having multiple classes.

In this context to classifiy the an email whether it is spam or not, I shall employ gaussian Naive Bayes Classifier model and test different metrics on the model and observe it's performance. 

Gaussian Naive Bayes Classifier model shall be applied since the predictor data is continuos

### Recording experimental design

The following are the steps taken to achieve the set objectives:
1. Importing the relevant libraries
2. Loading dataset
3. Previewing the dataset
4. Tidying the dataset- this will entail steps such as checking validity of the attributes in the dataset, check for any missing values in the dataset, check and deal with duplicates appropiately in the dataset
5. Implementing the solution - Here four different models shall be created using 80%, 70%, 60% and Finally 20% of the data to train the model and observing the performance metrics of each classifier
6. Optimization of the model with the best performance metrics
7. Challenging the naives bayes model using a random forest classifier
8. Drawing conclusions
9. Follow up questions

### Data Relevance

SPAM E-MAIL DATABASE ATTRIBUTES (in .names format)

48 continuous real [0,100] attributes of type word_freq_WORD  = percentage of words in the e-mail that match WORD, i.e. 100 * (number of times the WORD appears in the e-mail) / total number of words in e-mail.  A "word" in this case is any 
string of alphanumeric characters bounded by non-alphanumeric characters or end-of-string.

6 continuous real [0,100] attributes of type char_freq_CHAR = percentage of characters in the e-mail that match CHAR, i.e. 100 * (number of CHAR occurences) / total characters in e-mail

1 continuous real [1,...] attribute of type capital_run_length_average
= average length of uninterrupted sequences of capital letters

1 continuous integer [1,...] attribute of type capital_run_length_longest
= length of longest uninterrupted sequence of capital letters

1 continuous integer [1,...] attribute of type capital_run_length_total = sum of length of uninterrupted sequences of capital letters = total number of capital letters in the e-mail

1 nominal {0,1} class attribute of type spam = denotes whether the e-mail was considered spam (1) or not (0), i.e. unsolicited commercial e-mail.  

1. 1, 0.    | spam, non-spam classes

2. word_freq_make:         continuous.
3. word_freq_address:      continuous.
4. word_freq_all:          continuous.
5. word_freq_3d:           continuous.
6. word_freq_our:          continuous.
7. word_freq_over:         continuous.
8. word_freq_remove:       continuous.
9. word_freq_internet:     continuous.
10. word_freq_order:        continuous.
11. word_freq_mail:         continuous.
12. word_freq_receive:      continuous.
13. word_freq_will:         continuous.
14. word_freq_people:       continuous.
15. word_freq_report:       continuous.
16. word_freq_addresses:    continuous.
17. word_freq_free:         continuous.
18. word_freq_business:     continuous.
19. word_freq_email:        continuous.
20. word_freq_you:          continuous.
21. word_freq_credit:       continuous.
22. word_freq_your:         continuous.
23. word_freq_font:         continuous.
24. word_freq_000:          continuous.
25. word_freq_money:        continuous.
26. word_freq_hp:           continuous.
27. word_freq_hpl:          continuous.
28. word_freq_george:       continuous.
29. word_freq_650:          continuous.
30. word_freq_lab:          continuous.
31. word_freq_labs:         continuous.
32. word_freq_telnet:       continuous.
33. word_freq_857:          continuous.
34. word_freq_data:         continuous.
35. word_freq_415:          continuous.
36. word_freq_85:           continuous.
37. word_freq_technology:   continuous.
38. word_freq_1999:         continuous.
39. word_freq_parts:        continuous.
40. word_freq_pm:           continuous.
41. word_freq_direct:       continuous.
42. word_freq_cs:           continuous.
43. word_freq_meeting:      continuous.
44. word_freq_original:     continuous.
45. word_freq_project:      continuous.
46. word_freq_re:           continuous.
47. word_freq_edu:          continuous.
48. word_freq_table:        continuous.
49. word_freq_conference:   continuous.
50. char_freq_;:            continuous.
51. char_freq_(:            continuous.
52. char_freq_[:            continuous.
53. char_freq_!:            continuous.
54. char_freq_$:            continuous.
55. char_freq_#:            continuous.
56. capital_run_length_average: continuous.
57. capital_run_length_longest: continuous.
58. capital_run_length_total:   continuous.


## Importing libraries

In [85]:
# Importing pandas
import pandas as pd

# Importing numpy
import numpy as np

# Importing matplotlib
import matplotlib.pyplot as plt

# Importing seaborn
import seaborn as sns

# Importing datetime
import datetime

# Importing scipy.stats
import scipy.stats as stats

# Importing scipy
import scipy as sp

# Importing labelencoder from sklearn preprocessing
from sklearn.preprocessing import LabelEncoder

# Importing GaussianNB from sklearn naive bayes library
from sklearn.naive_bayes import GaussianNB

# Importing train test split from sklearn model selection library
from sklearn.model_selection import train_test_split

# Importing MinMaxScaler from sklearn preprocessing library
from sklearn.preprocessing import MinMaxScaler

## Loading dataset

In [29]:
# Loading the spam dataset
data =  pd.read_csv('/content/spambase.data', sep=",", header = None)


## Previewing dataset

In [30]:
# Previewing the first five records of the spam data frame
data.head()

0     1     2     3     4     5     6     7     8     9   ...    48  \
0 0.000 0.640 0.640 0.000 0.320 0.000 0.000 0.000 0.000 0.000  ... 0.000   
1 0.210 0.280 0.500 0.000 0.140 0.280 0.210 0.070 0.000 0.940  ... 0.000   
2 0.060 0.000 0.710 0.000 1.230 0.190 0.190 0.120 0.640 0.250  ... 0.010   
3 0.000 0.000 0.000 0.000 0.630 0.000 0.310 0.630 0.310 0.630  ... 0.000   
4 0.000 0.000 0.000 0.000 0.630 0.000 0.310 0.630 0.310 0.630  ... 0.000   

     49    50    51    52    53    54   55    56  57  
0 0.000 0.000 0.778 0.000 0.000 3.756   61   278   1  
1 0.132 0.000 0.372 0.180 0.048 5.114  101  1028   1  
2 0.143 0.000 0.276 0.184 0.010 9.821  485  2259   1  
3 0.137 0.000 0.137 0.000 0.000 3.537   40   191   1  
4 0.135 0.000 0.135 0.000 0.000 3.537   40   191   1  

[5 rows x 58 columns]

In [31]:
# Previewing last five records of the spam dataset
data.tail()

0     1     2     3     4     5     6     7     8     9   ...    48  \
4596 0.310 0.000 0.620 0.000 0.000 0.310 0.000 0.000 0.000 0.000  ... 0.000   
4597 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   
4598 0.300 0.000 0.300 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.102   
4599 0.960 0.000 0.000 0.000 0.320 0.000 0.000 0.000 0.000 0.000  ... 0.000   
4600 0.000 0.000 0.650 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   

        49    50    51    52    53    54  55   56  57  
4596 0.232 0.000 0.000 0.000 0.000 1.142   3   88   0  
4597 0.000 0.000 0.353 0.000 0.000 1.555   4   14   0  
4598 0.718 0.000 0.000 0.000 0.000 1.404   6  118   0  
4599 0.057 0.000 0.000 0.000 0.000 1.147   5   78   0  
4600 0.000 0.000 0.125 0.000 0.000 1.250   5   40   0  

[5 rows x 58 columns]

In [32]:
# Previewing column information of the attributes in the spam dataset
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4601 entries, 0 to 4600
Data columns (total 58 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       4601 non-null   float64
 1   1       4601 non-null   float64
 2   2       4601 non-null   float64
 3   3       4601 non-null   float64
 4   4       4601 non-null   float64
 5   5       4601 non-null   float64
 6   6       4601 non-null   float64
 7   7       4601 non-null   float64
 8   8       4601 non-null   float64
 9   9       4601 non-null   float64
 10  10      4601 non-null   float64
 11  11      4601 non-null   float64
 12  12      4601 non-null   float64
 13  13      4601 non-null   float64
 14  14      4601 non-null   float64
 15  15      4601 non-null   float64
 16  16      4601 non-null   float64
 17  17      4601 non-null   float64
 18  18      4601 non-null   float64
 19  19      4601 non-null   float64
 20  20      4601 non-null   float64
 21  21      4601 non-null   float64
 22  

All columns have appropiate data types

In [33]:
# Previewing the shape(rows and columns) of the training dataset

rows, columns = data.shape

# Printing out the number of rows and columns in the training dataset
print(f'The training dataset has {rows} rows, and {columns} columns')

The training dataset has 4601 rows, and 58 columns


## Tidying dataset

### Validity

In [34]:
# Checking the column names in the spam dataset
data.columns

Int64Index([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
            17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33,
            34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50,
            51, 52, 53, 54, 55, 56, 57],
           dtype='int64')

In [35]:
# Renaming column 57(the target variable) in the spam dataset appropiately
data.rename(columns = {57: 'spam or not spam'}, inplace = True)

In [36]:
# Previewing column names after renaming the target variable in spam dataset appropiately
data.columns

Index([                 0,                  1,                  2,
                        3,                  4,                  5,
                        6,                  7,                  8,
                        9,                 10,                 11,
                       12,                 13,                 14,
                       15,                 16,                 17,
                       18,                 19,                 20,
                       21,                 22,                 23,
                       24,                 25,                 26,
                       27,                 28,                 29,
                       30,                 31,                 32,
                       33,                 34,                 35,
                       36,                 37,                 38,
                       39,                 40,                 41,
                       42,                 43,                

Columns have been appropiate data types for the analysis. The predictor columns shall maintain their numerical naming but are defined in the data relevance section

### Missing values

In [37]:
# Checking for anly missing values in the spam dataset
data.isnull().any().sum()

0

The dataset does not have any missing values

### Duplicates

In [38]:
# Checking for any duplicate records in the spam dataset
print(f' The spam dataset has {data.duplicated().sum()} duplicate records')

 The spam dataset has 391 duplicate records


In [39]:
# Select duplicate rows except first occurrence based on all columns in the spam dataset
duplicates_df = data[data.duplicated()]

# Print duplicates dataframe first five records
duplicates_df.head()

0     1     2     3     4     5     6     7     8     9  ...    48  \
26  0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   
103 0.000 0.000 0.640 0.000 0.000 0.640 0.000 0.000 0.000 0.000  ... 0.094   
104 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.305   
105 0.000 0.000 0.640 0.000 0.000 0.640 0.000 0.000 0.000 0.000  ... 0.094   
106 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.305   

       49    50    51    52    53     54   55   56  spam or not spam  
26  0.196 0.000 0.392 0.196 0.000  5.466   22   82                 1  
103 0.189 0.284 0.662 0.000 0.000 10.068  131  292                 1  
104 0.611 0.000 1.529 0.000 0.000  5.500   22   66                 1  
105 0.189 0.284 0.662 0.000 0.000 10.068  131  292                 1  
106 0.611 0.000 1.529 0.000 0.000  5.500   22   66                 1  

[5 rows x 58 columns]

In [40]:
# Preview the last five records of the duplicates dataframe
duplicates_df.tail()

0     1     2     3     4     5     6     7     8     9  ...    48  \
4439 0.000 0.000 0.740 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.036   
4441 0.000 0.000 0.740 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.036   
4537 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   
4541 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   
4550 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000 0.000  ... 0.000   

        49    50    51    52    53    54  55   56  spam or not spam  
4439 0.147 0.000 0.147 0.000 0.000 2.587  55  282                 0  
4441 0.147 0.000 0.147 0.000 0.000 2.587  55  282                 0  
4537 0.000 0.000 0.000 0.000 0.000 1.000   1    6                 0  
4541 0.000 0.000 0.000 0.000 0.000 1.000   1    2                 0  
4550 0.000 0.000 0.000 0.000 0.000 2.000   3    4                 0  

[5 rows x 58 columns]

In [41]:
# Dropping duplicate records in the spam dataset
spam_df = data.drop_duplicates()

In [42]:
# Check for any remaining duplicate records in the spam dataset
spam_df.duplicated().sum()

0

In [43]:
# Preview records lost after dropping duplicates
print(f'The number of records lost after dropping duplicates was: {(len(data)- len(spam_df))}')

The number of records lost after dropping duplicates was: 391


## Exploratory data analysis

### Univariate analysis

In [44]:
# The numeric columns in the spam dataset will be analysed for mean, standard deviation and median values
# Describing data in spam dataset
spam_df.drop(columns = ['spam or not spam']).describe()

0        1        2        3        4        5        6        7   \
count 4210.000 4210.000 4210.000 4210.000 4210.000 4210.000 4210.000 4210.000   
mean     0.104    0.113    0.291    0.063    0.325    0.097    0.117    0.108   
std      0.300    0.454    0.516    1.352    0.688    0.276    0.397    0.410   
min      0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
25%      0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
50%      0.000    0.000    0.000    0.000    0.000    0.000    0.000    0.000   
75%      0.000    0.000    0.440    0.000    0.410    0.000    0.000    0.000   
max      4.540   14.280    5.100   42.810   10.000    5.880    7.270   11.110   

            8        9   ...       47       48       49       50       51  \
count 4210.000 4210.000  ... 4210.000 4210.000 4210.000 4210.000 4210.000   
mean     0.092    0.248  ...    0.035    0.040    0.144    0.017    0.281   
std      0.282    0.657  ...    0.299    0.253    0.274    0.106    0.843   
min      0.000    0.000  ...    0.000    0.000    0.000    0.000    0.000   
25%      0.000    0.000  ...    0.000    0.000    0.000    0.000    0.000   
50%      0.000    0.000  ...    0.000    0.000    0.073    0.000    0.016   
75%      0.000    0.190  ...    0.000    0.000    0.194    0.000    0.331   
max      5.260   18.180  ...   10.000    4.385    9.752    4.081   32.478   

            52       53       54       55        56  
count 4210.000 4210.000 4210.000 4210.000  4210.000  
mean     0.076    0.046    5.384   52.140   291.182  
std      0.240    0.436   33.147  199.582   618.655  
min      0.000    0.000    1.000    1.000     1.000  
25%      0.000    0.000    1.627    7.000    40.000  
50%      0.000    0.000    2.297   15.000   101.500  
75%      0.053    0.000    3.707   44.000   273.750  
max      6.003   19.829 1102.500 9989.000 15841.000  

[8 rows x 57 columns]

In [45]:
# Checking the distribution of the predictors in the spam dataset
# Skew
spam_df.drop(columns = ['spam or not spam']).skew(axis=0, skipna=True, level=None, numeric_only=None)

0     5.870
1    13.465
2     2.981
3    27.210
4     4.697
5     6.123
6     6.809
7     9.734
8     5.284
9     8.569
10    5.078
11    2.828
12    6.890
13   11.564
14    7.513
15    9.388
16    5.568
17    5.454
18    1.524
19   15.430
20    2.246
21    9.542
22    5.860
23   15.557
24    5.639
25    6.297
26    9.482
27    6.518
28   11.684
29    6.381
30   13.515
31   10.432
32   13.069
33   10.340
34   16.002
35    7.359
36    5.127
37   27.079
38   11.762
39    9.102
40   11.469
41    9.185
42    7.408
43   18.197
44    8.918
45   10.077
46   19.298
47   18.872
48   13.356
49   14.003
50   21.272
51   18.337
52   10.609
53   31.500
54   22.761
55   31.205
56    8.802
dtype: float64

All attributes had positive skews. This implied the data in each attribute was not normally distributed bust skewed to the right

In [46]:
# Checking the distribution of the predictors in the spam dataset
# kurtosis
spam_df.drop(columns = ['spam or not spam']).kurtosis(axis=0, skipna=True, level=None, numeric_only=None)

0      54.351
1     306.686
2      12.941
3     788.370
4      37.065
5      71.603
6      76.382
7     167.550
8      48.011
9     162.539
10     34.656
11     12.096
12     82.596
13    219.645
14     69.494
15    157.692
16     43.514
17     48.157
18      4.998
19    424.686
20      8.094
21     99.666
22     49.120
23    340.176
24     43.182
25     63.073
26    111.100
27     58.669
28    189.409
29     49.105
30    302.370
31    128.296
32    290.736
33    126.437
34    491.542
35     79.988
36     40.888
37    836.229
38    205.959
39    101.120
40    166.262
41    108.995
42     73.795
43    449.018
44    122.068
45    150.120
46    431.309
47    492.147
48    200.450
49    404.370
50    664.620
51    579.438
52    188.463
53   1238.321
54    614.400
55   1471.216
56    146.108
dtype: float64

All the predictors had a positive kurtosis. This implied that these variables had more tails than the normal distribution.

### Bivariate analysis

In [47]:
# The correlations among the features in the spam dataset to see which attributes have a high correlation amongst each other
corr = spam_df.corr()

# Correlations
corr

0      1      2      3      4      5      6      7  \
0                 1.000  0.034  0.063  0.005  0.022  0.055  0.011 -0.004   
1                 0.034  1.000  0.028 -0.009  0.037  0.006  0.077  0.014   
2                 0.063  0.028  1.000 -0.020  0.066  0.067  0.029  0.007   
3                 0.005 -0.009 -0.020  1.000  0.000 -0.009  0.014  0.004   
4                 0.022  0.037  0.066  0.000  1.000  0.041  0.136  0.023   
5                 0.055  0.006  0.067 -0.009  0.041  1.000  0.047  0.080   
6                 0.011  0.077  0.029  0.014  0.136  0.047  1.000  0.034   
7                -0.004  0.014  0.007  0.004  0.023  0.080  0.034  1.000   
8                 0.104  0.053  0.078 -0.001  0.014  0.097  0.049  0.107   
9                 0.041  0.184  0.025 -0.004  0.027  0.010  0.056  0.079   
10                0.063  0.061  0.056 -0.013  0.079  0.057  0.187  0.153   
11                0.076  0.026  0.066 -0.019  0.055  0.002 -0.008 -0.007   
12                0.062  0.016  0.036 -0.013  0.021  0.067  0.003  0.024   
13                0.032  0.012  0.005  0.003  0.002  0.008 -0.024  0.009   
14                0.022  0.062  0.105  0.005  0.048  0.140  0.037  0.072   
15                0.065  0.047  0.062  0.006  0.079  0.017  0.136  0.052   
16                0.081  0.022  0.030 -0.003  0.138  0.058  0.190  0.219   
17                0.052  0.179  0.111  0.017  0.051  0.059  0.115  0.026   
18                0.121  0.056  0.127 -0.010  0.083  0.083  0.103  0.017   
19                0.021 -0.007  0.031 -0.005  0.028  0.058  0.050  0.116   
20                0.121  0.118  0.148  0.005  0.143  0.106  0.144  0.170   
21               -0.026  0.001 -0.039  0.031 -0.023  0.007 -0.003 -0.017   
22                0.126  0.007  0.107  0.015  0.062  0.182  0.068  0.087   
23                0.206  0.061  0.031  0.026  0.000  0.059  0.035  0.036   
24               -0.075 -0.054 -0.093 -0.016 -0.077 -0.087 -0.095 -0.056   
25               -0.063 -0.045 -0.066 -0.014 -0.080 -0.090 -0.085 -0.044   
26               -0.048 -0.011 -0.081 -0.010 -0.062 -0.051 -0.060 -0.050   
27               -0.050 -0.033 -0.051 -0.011 -0.064 -0.068 -0.070 -0.052   
28               -0.042 -0.026 -0.058 -0.008  0.040 -0.049 -0.050 -0.038   
29               -0.054 -0.029 -0.031 -0.011 -0.054 -0.048 -0.061 -0.045   
30               -0.039 -0.016 -0.037 -0.008 -0.042 -0.046 -0.048 -0.036   
31               -0.032  0.029 -0.064 -0.007 -0.024 -0.037 -0.043 -0.035   
32               -0.043 -0.033 -0.058 -0.008 -0.035 -0.035 -0.043 -0.040   
33               -0.026  0.026 -0.063 -0.007 -0.025 -0.037 -0.043 -0.036   
34               -0.046 -0.026 -0.049 -0.006 -0.051 -0.055 -0.055 -0.036   
35               -0.057 -0.028 -0.048 -0.006 -0.052 -0.054 -0.057 -0.037   
36               -0.059  0.004 -0.074 -0.007 -0.078 -0.061 -0.056 -0.022   
37               -0.009 -0.013  0.032 -0.003  0.133 -0.018 -0.015 -0.013   
38               -0.011 -0.016 -0.017 -0.005 -0.046 -0.049 -0.049 -0.032   
39               -0.035  0.002 -0.046 -0.008 -0.018 -0.028 -0.020 -0.003   
40               -0.005 -0.019 -0.031 -0.006 -0.052 -0.031 -0.036 -0.000   
41               -0.027 -0.036 -0.008 -0.008  0.108 -0.057 -0.052 -0.045   
42               -0.026  0.038 -0.053 -0.009 -0.055 -0.038 -0.053 -0.005   
43               -0.023 -0.029 -0.057 -0.006  0.003 -0.030 -0.036 -0.031   
44               -0.038  0.020 -0.057 -0.013 -0.049 -0.057 -0.057 -0.007   
45               -0.034 -0.026 -0.061 -0.009 -0.083 -0.034 -0.059 -0.039   
46               -0.001 -0.012  0.027 -0.003 -0.028 -0.015 -0.018 -0.007   
47               -0.019 -0.021 -0.029 -0.002 -0.035 -0.033 -0.033 -0.022   
48               -0.027  0.015 -0.037 -0.000 -0.035 -0.022 -0.034 -0.029   
49               -0.020 -0.028 -0.025 -0.011 -0.055 -0.012 -0.061 -0.043   
50               -0.035 -0.019 -0.036 -0.008 -0.028 -0.017 -0.029 -0.021   
51                0.059  0.032  0.097 -0.004  0.019  0.058  0.051  0.0

The attributes in this data all seem to be loosely correlated. To confirm the abscence of multicolinearity in this dataset the VIF scores of the attributes shall be imputed.


In [48]:
# python suppress exponential notation
pd.set_option('display.float_format', lambda x: '%.3f' % x)

# VIF scores
pd.DataFrame(np.linalg.inv(corr.values), index = corr.index, columns=corr.columns)


0      1      2      3      4      5      6      7  \
0                 1.101 -0.004 -0.016 -0.002  0.007 -0.014  0.033  0.048   
1                -0.004  1.108  0.004  0.016 -0.010  0.016 -0.037  0.016   
2                -0.016  0.004  1.097  0.027 -0.011 -0.016  0.038  0.034   
3                -0.002  0.016  0.027  1.009  0.010  0.021  0.002  0.002   
4                 0.007 -0.010 -0.011  0.010  1.139  0.001 -0.046  0.041   
5                -0.014  0.016 -0.016  0.021  0.001  1.087  0.015 -0.038   
6                 0.033 -0.037  0.038  0.002 -0.046  0.015  1.196  0.074   
7                 0.048  0.016  0.034  0.002  0.041 -0.038  0.074  1.119   
8                -0.062 -0.006 -0.021  0.012  0.049 -0.038  0.022 -0.029   
9                -0.005 -0.189  0.024  0.003  0.005  0.041 -0.003 -0.039   
10               -0.003  0.003  0.006  0.024 -0.002  0.012 -0.121 -0.091   
11               -0.069 -0.028 -0.050  0.014 -0.037  0.003  0.020  0.020   
12               -0.020  0.011 -0.004  0.020 -0.002 -0.027  0.036 -0.002   
13               -0.008 -0.011  0.010 -0.001  0.001  0.010  0.031  0.009   
14                0.054  0.007 -0.058  0.006 -0.028 -0.074  0.017 -0.030   
15               -0.034 -0.012 -0.020  0.012 -0.010  0.043 -0.035 -0.005   
16               -0.054  0.018  0.030  0.009 -0.096  0.010 -0.134 -0.176   
17               -0.025 -0.195 -0.060 -0.015 -0.002 -0.004 -0.059  0.022   
18               -0.048  0.022 -0.068  0.021 -0.022 -0.017 -0.008  0.076   
19                0.017  0.029  0.001  0.014  0.030 -0.021  0.036 -0.044   
20               -0.038 -0.066 -0.079  0.009 -0.053 -0.011  0.018 -0.098   
21                0.020  0.014  0.047 -0.023  0.044  0.008  0.041  0.034   
22               -0.109  0.045 -0.042 -0.004  0.010 -0.103  0.030 -0.017   
23               -0.183 -0.026  0.033 -0.021  0.057 -0.011  0.034  0.002   
24                0.019  0.025  0.044  0.006 -0.001  0.016  0.013  0.020   
25                0.007  0.017  0.000  0.004  0.030  0.031 -0.005 -0.018   
26                0.018 -0.003  0.047  0.005  0.029  0.010  0.015  0.024   
27                0.000  0.028  0.009  0.004  0.017  0.025  0.022  0.020   
28                0.025  0.024  0.050 -0.003 -0.032  0.003 -0.003 -0.006   
29                0.019  0.048 -0.060 -0.000  0.002 -0.023  0.007  0.009   
30               -0.004  0.030 -0.030  0.002  0.015  0.007  0.006  0.002   
31                0.465 -0.522 -0.080 -0.030 -0.129 -0.103 -0.111 -0.033   
32                0.019  0.012  0.036  0.006  0.006  0.009  0.001  0.018   
33               -0.541  0.284  0.115  0.013  0.059  0.052  0.092  0.065   
34                0.007  0.005  0.015 -0.003  0.008  0.011 -0.006 -0.011   
35                0.027  0.050 -0.018  0.002  0.022  0.009  0.019 -0.003   
36                0.029 -0.026  0.020 -0.004  0.008  0.005 -0.019 -0.009   
37                0.001  0.011 -0.035 -0.001 -0.141  0.007  0.012  0.000   
38               -0.016  0.016 -0.016 -0.005  0.013  0.020  0.006  0.009   
39                0.039  0.075  0.027  0.011  0.021  0.041 -0.015 -0.033   
40               -0.024  0.006 -0.001  0.000  0.008  0.009 -0.006 -0.027   
41               -0.001  0.007 -0.034  0.002 -0.119  0.025  0.014  0.013   
42               -0.013 -0.043  0.016  0.002  0.015 -0.001  0.009 -0.021   
43                0.007  0.015  0.044  0.004 -0.029  0.010  0.010  0.015   
44                0.014 -0.042  0.036  0.005  0.011  0.031  0.010 -0.042   
45                0.017  0.012  0.040  0.006  0.045  0.007  0.017  0.019   
46                0.001  0.015 -0.032  0.001  0.022  0.007 -0.001 -0.001   
47                0.006  0.013  0.015 -0.000  0.016  0.015  0.005  0.005   
48                0.011 -0.029  0.004  0.009 -0.008  0.005 -0.007 -0.001   
49               -0.008  0.044 -0.012  0.024  0.023 -0.014  0.032  0.021   
50                0.022  0.017  0.012  0.006  0.000 -0.000  0.003  0.014   
51               -0.020 -0.004 -0.058  0.014  0.047 -0.012  0.038  0.0

The VIF score of all predictor variables are way below five. This affirms that the spam dataset does not suffer from multi colinearity

### Dimension reduction

The spam dataset has numerous predictor attributes (58). No dimesion reduction shall be applied since Naives bayes classifier is a powerful algorithm which works well with large datasets and sparse matrices.

## Implementing the solution

To predict whether an email is spam or not, the spam dataframe shall be split into the a train and test set. A Gaussian Naive Bayes Classifier algorithim shall be used since the predictors are continuos in nature. The predictors are assumed to be normally distributed.

The train set shall be be used to train the Gaussian Naive Bayes Classifier model using first 80% of data as train 20% as test, the 70% of data as train and 30% as test and finally 60% of data as train and 40% as test

The model shall be optimized and the metrics of optimized model analysed


### Model 1

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 80% of the data the test set is going to take up 20%

In [49]:
# Splitting the spam dataframe to get the target variables and the feature variables
# Feature variables
X = spam_df.drop(columns = ['spam or not spam'], axis =1)

# Target varibale
y = spam_df['spam or not spam']

In [51]:
# Splitting the target and feature variables  to get 80% of the data used in training the 
# Model and 20% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.20)

In [53]:
# Defining the classifier model
clf1 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf1.fit(X_train, y_train) 

In [54]:
# Making predictions from the test set of the spam dataset
#
pred = clf1.predict(X_test)

In [55]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[364 137]
 [ 12 329]]


Confusion matrix interpretation:

1. True positive = 364
2. True Negative = 329
3. False positive = 12
4. False negative = 137

In [56]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.73      0.83       501
           1       0.71      0.96      0.82       341

    accuracy                           0.82       842
   macro avg       0.84      0.85      0.82       842
weighted avg       0.86      0.82      0.82       842



The model (Model 1) had an accuracy score of 0.82

Using 80% of the data for the training set, the model (model 1) performed very well. It corretly predicted none spam emails 97/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 67 out 100 times correctly.

Model1 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

### Model 2

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 70% of the data the test set is going to take up 30%

In [57]:
# Splitting the target and feature variables  to get 70% of the data used in training the 
# Model and 30% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.30)

In [58]:
# Defining the classifier model
clf2 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf2.fit(X_train, y_train) 

In [59]:
# Making predictions from the test set of the spam dataset
#
pred = clf2.predict(X_test)

In [60]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[575 181]
 [ 26 481]]


Confusion matrix interpretation:

1. True positive = 575
2. True Negative = 481
3. False positive = 26
4. False negative = 181

In [61]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.96      0.76      0.85       756
           1       0.73      0.95      0.82       507

    accuracy                           0.84      1263
   macro avg       0.84      0.85      0.84      1263
weighted avg       0.86      0.84      0.84      1263



The model (Model 2) had an accuracy score of 0.84

Using 70% of the data for the training set, the model (model 2) performed very well. It corretly predicted none spam emails 96/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 69 out 100 times correctly.

Model2 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

However the model(model 2) generally produced better results relative to the first model(model 1)

### Model 3

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 60% of the data the test set is going to take up 40%

In [62]:
# Splitting the target and feature variables  to get 60% of the data used in training the 
# Model and 40% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.40)

In [63]:
# Defining the classifier model
clf3 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf3.fit(X_train, y_train) 

In [64]:
# Making predictions from the test set of the spam dataset
#
pred = clf2.predict(X_test)

In [65]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[771 248]
 [ 27 638]]


Confusion matrix interpretation:

1. True positive = 771
2. True Negative = 638
3. False positive = 27
4. False negative = 248

In [66]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.97      0.76      0.85      1019
           1       0.72      0.96      0.82       665

    accuracy                           0.84      1684
   macro avg       0.84      0.86      0.84      1684
weighted avg       0.87      0.84      0.84      1684



The model (Model 3) had an accuracy score of 0.84

Using 60% of the data for the training set, the model (model 2) performed very well. It corretly predicted none spam emails 97/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 73 out 100 times correctly.

Model3 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

Model 3 has similar performance metrics to model 2 but performs slightly better than model 1

### Model 4

This model shall employ a gaussian Naive Bayes Classifier model. The training set shall take up 20% of the data the test set is going to take up 80%.

This is a radical approach to challenge the finidngs of the prior three models. The train set is reduced significantlya and its performance metrics observed

In [72]:
# Splitting the target and feature variables  to get 60% of the data used in training the 
# Model and 40% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.80)

In [73]:
# Defining the classifier model
clf3 = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf3.fit(X_train, y_train) 

In [74]:
# Making predictions from the test set of the spam dataset
#
pred = clf2.predict(X_test)

In [75]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[1538  496]
 [  65 1269]]


Confusion matrix interpretation:

1. True positive = 1538
2. True Negative = 1269
3. False positive = 65
4. False negative = 496

In [76]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.96      0.76      0.85      2034
           1       0.72      0.95      0.82      1334

    accuracy                           0.83      3368
   macro avg       0.84      0.85      0.83      3368
weighted avg       0.86      0.83      0.84      3368



The model (Model 3) had an accuracy score of 0.83

Using 20% of the data for the training set, the model (model 2) performed very well. It corretly predicted none spam emails 96/100 times correctly. 

However it performed poorly in predicting spam emails. It correctly predicted spam emails 73 out 100 times correctly.

Model3 however had a poor recall for none spam recall but a had a strong recall for spam emails

From the F1 score, it can be concluded that the model performed performance in predicting spam and none spam emails was almost similar

Model 4 while vastly reducing the size of the training set did not significantly alter the performance metrictics of the gaussian naives bayes model

## Optimization

### Normalization of predictors

In creating the three models, I assumed the predictor data had a gaussin distribution. However, from the skew compuation most of the data was skewed to the right and not normally distributed

The data shall be normalized using a min max scaler. Since it has been determined earlier that the size of the train set did not significantly affect model metrics we shall use randomly use 65% of the data as train 

In [86]:
# Splitting the target and feature variables  to get 60% of the data used in training the 
# Model and 40% of the data to tests the model performance

# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.35)

In [87]:
# Normalization of the predictor data
# The feature variables are normalized to ensure that they are all in a normally distributed

# Specifying the normalizer
norm = MinMaxScaler()

# Fitting the X train data to the scaler
norm.fit(X_train)

MinMaxScaler()

In [88]:
# Normalization of the x train data
X_train = norm.transform(X_train)

# Normalization  of th x_test data
X_test = norm.transform(X_test)

In [89]:
# Defining the classifier model
clf = GaussianNB()

# Fitting the X_train and Y_train sets into the classifier model
model = clf.fit(X_train, y_train) 

In [90]:
# Making predictions from the test set of the spam dataset
#
pred = clf.predict(X_test)

In [91]:
# Evaluating the alghorithim
# The metrcs to evaluate the algorithim will be the confusion matrix, f1_score, recall, and precision

# Importing confusion matrix from sklearn metrics library
from sklearn.metrics import confusion_matrix

# Confusion matrix
print(confusion_matrix(y_test, pred))

[[635 239]
 [ 24 576]]


Confusion matrix interpretation:

1. True positive = 635
2. True Negative = 576
3. False positive = 24
4. False negative = 239

In [92]:
# Importing classification report from sklearn metrics library
from sklearn.metrics import classification_report

# Printing out the classification report of the alghorithim
print(classification_report(y_test, pred))

              precision    recall  f1-score   support

           0       0.96      0.73      0.83       874
           1       0.71      0.96      0.81       600

    accuracy                           0.82      1474
   macro avg       0.84      0.84      0.82      1474
weighted avg       0.86      0.82      0.82      1474



With normalizetion the performance metrics of the model remained fairly the same

## Conclusion

In conclusion, in a dataset with many records, the size of the train set does not affect the models performances significantly.

Even after normalization, the performance metrics of the model did not result in any significant change. It hence can be concluded that the distribution of the predictors does not affect the performance of a the gaussian naive bayes classifier model significantky

## Follow up questions

a). Do we have the right data?

For this machine learbung problem we have the right data

b) Is the data relevant to the study?

The data was relevant in aiding us achieving the aims of our modelling